skada/utils.py

```diff
@@ -0,0 +1,247 @@
+        if y_type == 'classification':
+            mask = (y == _DEFAULT_MASKED_TARGET_CLASSIFICATION_LABEL)
+        else:
+            mask = (np.isnan(y))
+        sample_domain[mask] = _DEFAULT_TARGET_DOMAIN_LABEL
+
+    source_idx = extract_source_indices(sample_domain)
+
+    # xxx(okachaiev): this needs to be re-written to accommodate for a
+    # a new domain labeling convention without "intersections"
+    n_sources = np.unique(sample_domain[source_idx]).shape[0]
+    n_targets = np.unique(sample_domain[~source_idx]).shape[0]
+
+    if not allow_source and n_sources > 0:
+        raise ValueError(f"Number of sources provided is {n_sources} "
+                         "and 'allow_source' is set to False")
+    if not allow_target and n_targets > 0:
+        raise ValueError(f"Number of targets provided is {n_targets} "
+                         "and 'allow_target' is set to False")
+    if not allow_multi_source and n_sources > 1:
+        raise ValueError(f"Number of sources provided is {n_sources} "
+                         "and 'allow_multi_source' is set to False")
+    if not allow_multi_target and n_sources > 1:
+        raise ValueError(f"Number of targets provided is {n_targets} "
+                         "and 'allow_multi_target' is set to False")
+
+    return X, y, sample_domain
+
+
+# xxx(okachaiev): code duplication, just for testing
+def check_X_domain(
+    X,
+    sample_domain,
+    *,
+    allow_domains: Optional[Set[int]] = None,
+    allow_source: bool = True,
+    allow_multi_source: bool = True,
+    allow_target: bool = True,
+    allow_multi_target: bool = True,
+    # xxx(okachaiev): most likely this needs to be removed as it doesn't fit new API
+    allow_auto_sample_domain: bool = True,
+):
+    """
+    Input validation for domain adaptation (DA) estimator.
+    If we work in single-source and single target mode, return source and target
+    separately to avoid additional scan for 'sample_domain' array.
+
+    Parameters:
+    ----------
+    X : array-like of shape (n_samples, n_features)
+        Input features.
+    sample_domain : array-like of shape (n_samples,)
+        Domain labels for each sample.
+    allow_domains : set of int, optional (default=None)
+        Set of allowed domain labels. If provided, only these domain labels are allowed.
+    allow_source : bool, optional (default=True)
+        Allow the presence of source domains.
+    allow_multi_source : bool, optional (default=True)
+        Allow multiple source domains.
+    allow_target : bool, optional (default=True)
+        Allow the presence of target domains.
+    allow_multi_target : bool, optional (default=True)
+        Allow multiple target domains.
+    allow_auto_sample_domain : bool, optional (default=True)
+        Allow automatic generation of sample_domain if not provided.
+
+    Returns:
+    ----------
+    X : array
+        Combined input features for source and target domains.
```

Comment:
Should it be just 'input features' the same way it is marked in the parameters section?